x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/event/parser/event/AnalyticsEventPageView.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.application.analytics.event.parser.event;
+
+import org.elasticsearch.xcontent.ObjectParser;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xpack.application.analytics.event.AnalyticsEvent;
+import org.elasticsearch.xpack.application.analytics.event.parser.field.AnalyticsEventDocumentField;
+import org.elasticsearch.xpack.application.analytics.event.parser.field.AnalyticsEventPageField;
+import org.elasticsearch.xpack.application.analytics.event.parser.field.AnalyticsEventSessionField;
+import org.elasticsearch.xpack.application.analytics.event.parser.field.AnalyticsEventUserField;
+
+import java.io.IOException;
+
+import static org.elasticsearch.xpack.application.analytics.event.parser.field.AnalyticsEventDocumentField.DOCUMENT_FIELD;
+import static org.elasticsearch.xpack.application.analytics.event.parser.field.AnalyticsEventPageField.PAGE_FIELD;
+import static org.elasticsearch.xpack.application.analytics.event.parser.field.AnalyticsEventSessionField.SESSION_FIELD;
+import static org.elasticsearch.xpack.application.analytics.event.parser.field.AnalyticsEventUserField.USER_FIELD;
+
+public class AnalyticsEventPageView {
```

Comment:
Naming: `PageViewAnalyticsEvent`, `SearchAnalyticsEvent`, `SearchClickAnalyticsEvent` is what I would expect in an `AnalyticsEvent` hierarchy, with the main type at the end.

Why not make it an actual hierarchy, with classes extending `AnalyticsEvent`? That would allow to define a `getParser()` and `fromXContent()` as abstract methods in the superclass...

Reply:
I tried this approach first but it was very cumbersome and the codebase was twice bigger.

Instead I choose to create an AnalyticsEvent which does not have to know about the content of the payload and build separate parsers for each event types. With this approach the code has become smaller but also easier to understand with less conditionals. In the end it has been a lot easier to write the tests.

PS: In general the parser code is hard to inherits because it is static methods. This is one more argument that make inheritance even less interesting in this case.

Reply:
Got it, thanks for clarifying!